File: Medium/Prime Pair with Target Sum/prime-pair-with-target-sum.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

class Array {
  public:
    template <class T>
    static void input(vector<T> &A, int n) {
        for (int i = 0; i < n; i++) {
            scanf("%d ", &A[i]);
        }
    }

    template <class T>
    static void print(vector<T> &A) {
        for (int i = 0; i < A.size(); i++) {
            cout << A[i] << " ";
        }
        cout << endl;
    }
};


// } Driver Code Ends

class Solution {
  public:
   bool isPrime(int n){
      for(int i=2;i<=sqrt(n);i++){
          if(n%i==0)return false;
      }
      return true;
  }
    vector<int> getPrimes(int n) {
        vector<int>result(2,-1);
        int s=2,e=n-1;
        while(s<=e){
            if(s+e==n){
                if(isPrime(s)&&isPrime(e)){
                    if(result[0]!=-1&&result[0]<s){
                        s++;
                        e--;
                        continue;
                    }
                    result[0]=s;
                    result[1]=e;  
                }
                s++;
                e--;
            }
            
            else if(s+e<n)s++;
            else e--;
            
            
        }
        return result;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    scanf("%d ", &t);
    while (t--) {

        int n;
        scanf("%d", &n);

        Solution obj;
        vector<int> res = obj.getPrimes(n);

        Array::print(res);
    }
}

// } Driver Code Ends

File: Difficulty: Medium/Prime Pair with Target Sum/prime-pair-with-target-sum.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;
import java.util.ArrayList;

class IntArray {
    public static int[] input(BufferedReader br, int n) throws IOException {
        String[] s = br.readLine().trim().split(" ");
        int[] a = new int[n];
        for (int i = 0; i < n; i++) a[i] = Integer.parseInt(s[i]);

        return a;
    }

    public static void print(int[] a) {
        for (int e : a) System.out.print(e + " ");
        System.out.println();
    }

    public static void print(ArrayList<Integer> a) {
        for (int e : a) System.out.print(e + " ");
        System.out.println();
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while (t-- > 0) {

            int n;
            n = Integer.parseInt(br.readLine());

            Solution obj = new Solution();
            ArrayList<Integer> res = obj.getPrimes(n);

            IntArray.print(res);
        }
    }
}

// } Driver Code Ends



class Solution {
    public static ArrayList<Integer> getPrimes(int n) {
        ArrayList<Integer> result = new ArrayList<>();

        // Generate all prime numbers up to n using the Sieve of Eratosthenes
        boolean[] isPrime = new boolean[n + 1];
        Arrays.fill(isPrime, true);
        isPrime[0] = isPrime[1] = false; // 0 and 1 are not primes

        for (int i = 2; i * i <= n; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        // Check pairs of primes to find the required pair
        for (int i = 2; i <= n / 2; i++) {
            if (isPrime[i] && isPrime[n - i]) {
                result.add(i);
                result.add(n - i);
                return result; // Return the first pair found
            }
        }

        // If no pair is found, return [-1, -1]
        result.add(-1);
        result.add(-1);
        return result;
    }
}
